### DISCUSSION TITLE
Temporary File in Setup MSI

### DISCUSSION THREAD

**Discussion by kplatter** [oleg-shilo/wixsharp#1476] [General]
Is there a way to include a file in the setup msi but not install it, and then use it in a CA or AfterInstall script?

I have a virgin INI file that if an INI file does not exist in (several locations that are checked by c#) on a users system I would like to place the renamed temporary file onto the users machine.

**Comment by oleg-shilo** [comment]
```C#
var project = new ManagedProject(... 
                  new Binary(new Id("my_ini"), @".\dir1\dir2\settings.ini"),
                  . . .
          
static void project_AfterInstall(SetupEventArgs e)
{
    byte[] ini_data = e.Session.GetEmbeddedData("my_ini");          
    . . . 
```

**Comment by kplatter** [comment]
Thanks oleg for the response, but calling e.Session.GetEmbeddedData causes error: "Cannot access session details from a non-immediate custom action" and GetEmbeddedData returns null.  I am calling in project_AfterInstall

**Comment by oleg-shilo** [comment]
Ah, of course. I forgot that you do it after the installation when the session is terminated. :o)
You can do a few things in this case:
- store the file in some known to you temp location and access it from `AfterInstall`
- Alternatively, you can completely ignore MSI and add the file as a resource to your app (as with any .NET application). Then it becomes a part of your assembly so you have no dependency on anything.

**Comment by kplatter** [comment]
:)  I was able to get it to work using e.Data so that it would be available after session was terminated, but I like your idea of putting it  as a resource in the app (much better in my opinion)

I added the file as a resource and had to set it's properties to "Embedded resource", then

I created a function:

``` C#
        public static string GetFromResources(string resourceName)
        {
            System.Reflection.Assembly assembly = System.Reflection.Assembly.GetExecutingAssembly();

            //MessageBox.Show(string.Join(",", assembly.GetManifestResourceNames()));

            using (Stream stream = assembly.GetManifestResourceStream("WixSharp.Resources." + resourceName))
            {
                using (var reader = new StreamReader(stream))
                {
                    return reader.ReadToEnd();
                }
            }
        }
```

**Reply by oleg-shilo** [reply]
Perfect. Even though e.Data was a special WixSharp trick to overcome WiX/MSI limitations, a simple generic .NET technique as embedding resources seems like an overlooked option.

**Comment by kplatter** [comment]
Glad you thought of it ;)
